### ISSUE TITLE
pl.DataFrame from a numpy array of dimension (n, 1) makes a dataframe of n rows of list

### ISSUE THREAD

**Issue by PierreAttard** [pola-rs/polars#12718]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import numpy as np
import polars as pl

pl.DataFrame(np.arange(4).reshape(-1, 1))
```

### Issue description

I think an issue has been introduced into the version 0.19.16 with the PR [!256](https://github.com/pola-rs/polars/pull/12672) .

I would like to get the same result with those two commands like in version 0.19.15:

```pycon
>>> pl.DataFrame(np.arange(4))
... ┌──────────┐
... │ column_0 │
... │ ---      │
... │ i32      │
... ╞══════════╡
... │ 0        │
... │ 1        │
... │ 2        │
... │ 3        │
... └──────────┘

>>> pl.DataFrame(np.arange(4).reshape(-1, 1))
... ┌──────────┐
... │ column_0 │
... │ ---      │
... │ i32      │
... ╞══════════╡
... │ 0        │
... │ 1        │
... │ 2        │
... │ 3        │
... └──────────┘

```

But I get those results in version 0.19.16 and 0.19.17:

```pycon
>>> pl.DataFrame(np.arange(4))
... ┌──────────┐
... │ column_0 │
... │ ---      │
... │ i32      │
... ╞══════════╡
... │ 0        │
... │ 1        │
... │ 2        │
... │ 3        │
... └──────────┘

>>> pl.DataFrame(np.arange(4).reshape(-1, 1))
... ┌───────────┐
... │ column_0  │
... │ ---       │
... │ list[i32] │
... ╞═══════════╡
... │ [0]       │
... │ [1]       │
... │ [2]       │
... │ [3]       │
... └───────────┘

```

Maybe is it the right behaviour but I think this is strange and not practical if you don't know in advance the number of columns you have.

I precise I have the same behaviour with `pl.from_numpy`

### Expected behavior

```pycon
>>> pl.DataFrame(np.arange(4))
... ┌──────────┐
... │ column_0 │
... │ ---      │
... │ i32      │
... ╞══════════╡
... │ 0        │
... │ 1        │
... │ 2        │
... │ 3        │
... └──────────┘

>>> pl.DataFrame(np.arange(4).reshape(-1, 1))
... ┌──────────┐
... │ column_0 │
... │ ---      │
... │ i32      │
... ╞══════════╡
... │ 0        │
... │ 1        │
... │ 2        │
... │ 3        │
... └──────────┘

```

### Installed versions

<details>

```
--------Version info---------
Polars:               0.19.17
Index type:           UInt32
Platform:             Windows-10-10.0.22621-SP0
Python:               3.10.11 (tags/v3.10.11:7d4cc5a, Apr  5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)]
----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          2.2.1
connectorx:           <not installed>
deltalake:            <not installed>
fsspec:               2023.6.0
gevent:               <not installed>
matplotlib:           3.7.2
numpy:                1.24.4
openpyxl:             <not installed>
pandas:               2.0.3
pyarrow:              11.0.0
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

**Comment by PierreAttard** [comment]
It seems it is here the issue

![image](https://github.com/pola-rs/polars/assets/11873372/39bee5f7-e9d9-4cd3-b8c6-decb001d53d5)

**Comment by PierreAttard** [comment]
For instance, on that very simplified example, you don't know the dimension of the input array. Whatever the value oh the dimension 1, you want a Polars DataFrame with the same number of columns. If not you must have a specific case if the input data has only one column.

```pycon
>>> def f(arr: np.ndarray, schema) -> pl.DataFrame:
...     return pl.DataFrame(arr, schema)
>>> f(np.zeros((2, 2)), ["a", "b"])
... ┌─────┬─────┐
... │ a   ┆ b   │
... │ --- ┆ --- │
... │ f64 ┆ f64 │
... ╞═════╪═════╡
... │ 0.0 ┆ 0.0 │
... │ 0.0 ┆ 0.0 │
... └─────┴─────┘
>>> f(np.zeros((2, 1)), ["a"])
... ┌──────┐
... │ a  │
... │ --- │
... │ f64 │
... ╞══════╡
... │ 0.0 │
... │ 0.0 │
... └──────┘
```

**Comment by Elvynzs** [comment]
I think the current "expected behavior" is very confusing, as you can have different return types for the same input type, depending on the number of columns.
Also for this reason `pl.from_numpy(arr).to_numpy()` fails if arr is a 2d array of 1 column.

As of polars 0.19.17, the current behavior of `from_numpy` is :

- `pl.from_numpy(np.zeros((2)))` : 1 column, `f64` dtype
- `pl.from_numpy(np.zeros((2,1)))` : 1 column, `list[f64]` dtype
- ` pl.from_numpy(np.zeros((2,2)))` : 2 columns, `f64` dtype


For consistency purposes, I think the default behavior of `from_numpy` with a 2d array input should always return a pl.Dataframe with f64 dtype.
If we really want the `list[f64]` dtype, we can always use `pl.Series(arr)` which will always use a `list[f64]` dtype for 2d numpy arrays.


However if we want to keep the current behavior, I think we should update the documentation of [from_numpy](https://pola-rs.github.io/polars/py-polars/html/reference/api/polars.from_numpy.html) to warn the users of this edge case.

**Comment by alexander-beedie** [comment]
Hmm... I'll take a look at the DataFrame case; for Series it seems clear that we are doing the right thing, but there may be a bit more of a grey area around DataFrame 🤔

**Comment by PierreAttard** [comment]
@alexander-beedie , I propose a PR.

**Comment by alexander-beedie** [comment]
Nice one; thanks for the heads-up and the fix 👍 
(Closed by #12746)
